### DISCUSSION TITLE
Update v6.6 - dev.5

### DISCUSSION THREAD

**Discussion by TBog** [TBog/TBLauncher#209] [Pre-Release]
* improve tag rename
* improve tag menu (item height)
* fix TextInputLayout colors
* add animation for custom icon dialog expanding
* update icon scale while dragging seek bar

<hr /><em>This discussion was created from the release <a href='https://github.com/TBog/TBLauncher/releases/tag/v6.6-dev.5'>Update v6.6 - dev.5</a>.</em>

**Comment by ltguillaume** [comment]
1. When renaming a tag, TBLauncher crashed:
```
android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: favorites.record (code 2067 SQLITE_CONSTRAINT_UNIQUE) 
at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method) at 
android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:831) at 
android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:756) at 
android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:66) at 
android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java:1704) at 
android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java:1650) at 
rocks.tbog.tblauncher.db.DBHelper.renameTag(DBHelper.java:481) at rocks.tbog.tblauncher.TagsHandler.renameTag(TagsHandler.java:439) at 
rocks.tbog.tblauncher.TagsManager.applyChanges(TagsManager.java:54) at 
rocks.tbog.tblauncher.ui.TagsManagerDialog.lambda$onViewCreated$0$TagsManagerDialog(TagsManagerDialog.java:42) at 
rocks.tbog.tblauncher.ui.-$$Lambda$TagsManagerDialog$Q8eGn_TNUVXQAYyFoWEozMkeOAQ.onButtonClick(Unknown Source:2) at 
rocks.tbog.tblauncher.ui.DialogFragment.onButtonClick(DialogFragment.java:89) at 
rocks.tbog.tblauncher.ui.DialogFragment.lambda$createButtonBar$0$DialogFragment(DialogFragment.java:194) at 
rocks.tbog.tblauncher.ui.-$$Lambda$DialogFragment$5EM98buW-1XOawim0G-Len2YqgM.onClick(Unknown Source:2) at 
android.view.View.performClick(View.java:7259) at 
com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119) at 
android.view.View.performClickInternal(View.java:7236) at android.view.View.access$3600(View.java:801) at 
android.view.View$PerformClick.run(View.java:27892) at android.os.Handler.handleCallback(Handler.java:883) at 
android.os.Handler.dispatchMessage(Handler.java:100) at android.os.Looper.loop(Looper.java:214) at 
android.app.ActivityThread.main(ActivityThread.java:7356) at java.lang.reflect.Method.invoke(Native Method) at 
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491) at 
com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)
```
What I did was rename the tag `Media` back to `media` (what it was before v6.6 had a tag menu). I now also see the old icon, as set for `media` before, so the previous renaming kept the icon for the original tag name in the database... don't think that's supposed to happen...
Also, the tag manager _still_ shows `Media` after the crash, even though the tags menu now shows what I renamed it to: `media` (thus with the wrong, old icon).

2. The icons in the tags menu now seem to be overlapping and cut off at the top/bottom edges:

![Screenshot_20211007-192339_TinyBit_launcher](https://user-images.githubusercontent.com/4777345/136433826-8ee55444-a775-4c56-b264-170b607b0812.png)

(Maybe consider showing the tags list in the result list instead?)

3. The tags order dialog still has a green border

![Screenshot_20211007-193915_TinyBit_launcher](https://user-images.githubusercontent.com/4777345/136435843-6ae264ae-7aa3-4e7f-96c7-a30d92a6a2e1.png)

4. The apk size almost doubled, what's the cause of this?

**Reply by TBog** [reply]
1. Seems like you have an entry in the database with media and Media. I'll add a fix for this but as a workaround at this time we can fix this using the backup feature. Export favorites and then generated files look for `tag://media` and if you find more then one, choose one and delete the other. Then import using _overwrite_
_Note: to delete an entry you must remove the whole `<favorite>` to `</favorite>` entry_

I'd also be interested in the file as well so I can better test my fix.

2. If I show the tag menu in the results I'll be weird to have you press a tag and the results change. I'll make you a version so you can see what I mean.
If you don't like the icons touching you can make the icons in the tag menu smaller.

3. I was not aware of the green border.

4. Maybe the constraint layout version change? I'll check the older versions and find when this happened. Thanks for noticing.

**Reply by ltguillaume** [reply]
> 
> 
>     1. Seems like you have an entry in the database with media and Media. I'll add a fix for this but as a workaround at this time we can fix this using the backup feature. Export favorites and then generated files look for `tag://media` and if you find more then one, choose one and delete the other. Then import using _overwrite_
>        _Note: to delete an entry you must remove the whole `<favorite>` to `</favorite>` entry_

Did that, but when choosing "Overwrite imported settings", I just get to see this:

![Screenshot_20211008-155046_TinyBit_launcher](https://user-images.githubusercontent.com/4777345/136569988-d50039e5-1c92-4541-88c1-c4fad41b7297.png)

Looks like the button on the left doesn't do anything and the one on the right just cancels the operation.
I checked if I did anything wrong when editing by backing up the favorites and trying to import the file unedited: same result.

> I'd also be interested in the file as well so I can better test my fix.

Exported file: [favorites.zip](https://github.com/TBog/TBLauncher/files/7311885/favorites.zip)

>     1. If I show the tag menu in the results I'll be weird to have you press a tag and the results change. I'll make you a version so you can see what I mean.

Really? I've had tags in the quick list and switching between tags was initially broken, but by now works nicely. Yeah, I'd love to see what you mean.

>     2. If you don't like the icons touching you can make the icons in the tag menu smaller.

That's quite user unfriendly, I'd say. But yeah I'd love to see how it looks as result list instead.

> 
>     3. I was not aware of the green border.

No worries. As you saw in the screenshot above, that dialog also uses the green border, but for the entire dialog.
 
>     4. Maybe the constraint layout version change? I'll check the older versions and find when this happened. Thanks for noticing.

Great, thanks!

**Reply by TBog** [reply]
I've looked at the file you provided and indeed there were two media tags with upper and lower case `M`. 

The import settings is indeed looking broken but the reason it didn't work was because I told you to `Overwrite` instead of `Set`.
I fixed the UI on the development branch and I'll try to add a "Done successfully" notification after the import.

If I decrease the initial icon size of the tags menu icons will you still think it's unfriendly?

**Reply by ltguillaume** [reply]
> I've looked at the file you provided and indeed there were two media tags with upper and lower case `M`.

Jup, that was the issue indeed.

> The import settings is indeed looking broken but the reason it didn't work was because I told you to `Overwrite` instead of `Set`.

You're right, using `Set` with an edited xml file, and exporting the favorites afterwards did show that the "duplicate" tags were removed.

> I fixed the UI on the development branch and I'll try to add a "Done successfully" notification after the import.

Great!

**Comment by ildar** [comment]
@TBog , I'm so happy with the latest versions so I cannot ask of anything
else.
Thank you!!!

**Reply by TBog** [reply]
Do you know if I can have a stable and development release on F-Droid?

**Reply by ltguillaume** [reply]
> Do you know if I can have a stable and development release on F-Droid?

This should be possible, yes. For a user to access them, they should turn on `Expert mode` and enable `Unstable updates`.

**Reply by TBog** [reply]
And what do I need to do so that F-Droid knows what is an unstable release?
Right now I only create a tag with the format `vX.Y` and F-Droid picks it up.
I see on the forum that there is an `-unstable` suffix I can add. Is this enough?

**Reply by ltguillaume** [reply]
> 
> 
> And what do I need to do so that F-Droid knows what is an unstable release? Right now I only create a tag with the format `vX.Y` and F-Droid picks it up. I see on the forum that there is an `-unstable` suffix I can add. Is this enough?

Use the older stable version/versioncode for `CurrentVersion` and `CurrentVersionCode` in the build metadata AFAIK: https://f-droid.org/docs/Build_Metadata_Reference/

It could be that tags containing `alpha`, `beta` or perhaps even `dev` are picked up as unstable. A project that uses `beta` in their tags, with them only showing up when `Unstable updates` is enabled, is DAVx5. You could have a look at their `fdroiddata` entry.

**Reply by TBog** [reply]
I'm still not sure what I need to do. TBLauncher was added to FDroid by a request for packaging: https://gitlab.com/fdroid/rfp/-/issues/1424
I see that the setting I have is: `UpdateCheckMode: Tags`

**Reply by ltguillaume** [reply]
> I'm still not sure what I need to do. TBLauncher was added to FDroid by a request for packaging: https://gitlab.com/fdroid/rfp/-/issues/1424 I see that the setting I have is: `UpdateCheckMode: Tags`

Agreed, did some searching myself twice now, but it remains unclear to me. If anyone knows, it's Izzy... https://gitlab.com/IzzySoft
